In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error, make_scorer, mean_absolute_error, r2_score
from sklearn.kernel_ridge import KernelRidge
from sklearn import linear_model
from sklearn import metrics
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.metrics import classification_report
import warnings


warnings.filterwarnings('ignore')

In [5]:
# uploading dataset of wine
heart = pd.read_csv('C:/Users/Moiz/Desktop/heart.csv') 
heart.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [6]:
X = heart.drop(columns=['target'])
print(X)

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  
0        0   0     1  
1        0   0     2  
2        2   0    

In [7]:
# using the targetted values for training-testing
Y = heart['target'].values
print(Y)

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0]


In [8]:
# To Split the dataset into training and the test set into 70% training and 30% test
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.3) 

# Regular Ridge

In [10]:
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression

lin_reg =LinearRegression()
mse=cross_val_score(lin_reg,X,Y,scoring='neg_mean_squared_error',cv=5)
mean_mse=np.mean(mse)
print(mean_mse)

-0.1726548441249148


In [11]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV
from sklearn.kernel_ridge import KernelRidge
from sklearn.model_selection import KFold



ridge=Ridge()
parameters ={'alpha':[1e-15,1e-10,1e-8,1e-3,1e-2,1,5,10,20,30,45,50,55,100,200]}
ridge_reg=GridSearchCV(ridge,parameters,scoring='neg_mean_squared_error',cv=5)
ridge_reg.fit(X,Y)

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=Ridge(alpha=1.0, copy_X=True, fit_intercept=True,
                             max_iter=None, normalize=False, random_state=None,
                             solver='auto', tol=0.001),
             iid='warn', n_jobs=None,
             param_grid={'alpha': [1e-15, 1e-10, 1e-08, 0.001, 0.01, 1, 5, 10,
                                   20, 30, 45, 50, 55, 100, 200]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring='neg_mean_squared_error', verbose=0)

In [12]:
print(ridge_reg.best_params_) #gives you best alpha parameter
print(ridge_reg.best_score_) #gives you best score which is near to zero

{'alpha': 5}
-0.17202770267770146


#### The accuracy we got is the aplha value which gives the best score which is near to zero

#### we got aplha parametre as 5 which has -0.17202770267770146

# Kernel Ridge Regressor

In [13]:
from sklearn.linear_model import RidgeClassifier
import warnings

warnings.filterwarnings('ignore')

In [14]:
# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [15]:
ridge_reg = RidgeClassifier(alpha=0.01) 

ridge_reg.fit(X_train, y_train)
y_pred1 = ridge_reg.predict(X_test)
ridge_reg100 = RidgeClassifier(alpha=100) #  comparison with alpha value
ridge_reg100.fit(X_train, y_train)
y_pred2 = ridge_reg100.predict(X_test)

Ridge_train_score = ridge_reg.score(X_train,y_train)*100
Ridge_test_score = ridge_reg.score(X_test, y_test)*100

Ridge_train_score100 = ridge_reg100.score(X_train,y_train)*100
Ridge_test_score100 = ridge_reg100.score(X_test, y_test)*100


print("Models Prediction Accuracy:\n")

print ("Score for Ridge regression train data with low alpha: {:.4f}".format(Ridge_train_score))
print ("Score for Ridge regression test data with low alpha: {:.4f}\n".format( Ridge_test_score))

print ("Score for Ridge regression with train data with high alpha: {:.4f}".format(Ridge_train_score100))
print ("Score for Ridge regression with test data with high alpha: {:.4f}".format(Ridge_test_score100))

Models Prediction Accuracy:

Score for Ridge regression train data with low alpha: 83.0189
Score for Ridge regression test data with low alpha: 85.7143

Score for Ridge regression with train data with high alpha: 83.4906
Score for Ridge regression with test data with high alpha: 85.7143


## SVM

# Linear SVM

In [16]:
from sklearn import svm
clf = svm.SVC(kernel='linear',gamma='auto', C=2)
clf.fit(X_train, y_train)

SVC(C=2, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto', kernel='linear',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [17]:
y_pred = clf.predict(X_test)

In [18]:
y_pred

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1,
       0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0,
       1, 1, 0], dtype=int64)

In [19]:
print("Accuracy:",metrics.accuracy_score(y_test, y_pred)*100)

Accuracy: 83.51648351648352


In [20]:
print('Linear:',classification_report(y_test, y_pred))

Linear:               precision    recall  f1-score   support

           0       0.82      0.76      0.79        37
           1       0.84      0.89      0.86        54

    accuracy                           0.84        91
   macro avg       0.83      0.82      0.83        91
weighted avg       0.83      0.84      0.83        91



# Polynomial SVM

In [21]:
clfp = svm.SVC(kernel='poly', gamma='auto', C=2)

In [22]:
clfp.fit(X_train, y_train)

SVC(C=2, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto', kernel='poly',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [23]:
y_pred = clfp.predict(X_test)
print("Accuracy:",metrics.accuracy_score(y_test, y_pred)*100)

Accuracy: 82.41758241758241


In [24]:
print('Polynomial:',classification_report(y_test, y_pred))

Polynomial:               precision    recall  f1-score   support

           0       0.82      0.73      0.77        37
           1       0.83      0.89      0.86        54

    accuracy                           0.82        91
   macro avg       0.82      0.81      0.81        91
weighted avg       0.82      0.82      0.82        91



# RBF Kernel

In [25]:
clfr = svm.SVC(kernel='rbf',gamma='auto', C=2)
clfr.fit(X_train, y_train)

SVC(C=2, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [26]:
y_pred = clfr.predict(X_test)

In [27]:
print("Accuracy:",metrics.accuracy_score(y_test, y_pred)*100)

Accuracy: 85.71428571428571


In [28]:
print('RBF:',classification_report(y_test, y_pred))

RBF:               precision    recall  f1-score   support

           0       0.85      0.78      0.82        37
           1       0.86      0.91      0.88        54

    accuracy                           0.86        91
   macro avg       0.86      0.85      0.85        91
weighted avg       0.86      0.86      0.86        91



### So when we compare, we observe that RBF kernel trick got the best accuracy for this dataset.